### PR TITLE
Added frozen string literal in all files

### PR DESCRIPTION
**Pull request by redtachyons**
Enable frozen string literal inn all files

- [x] Make all tests passing
- [x] Fix things which are not covered in tests
- [x] Readme fix
- [ ] Changelog

### CODE REVIEW

**Comment by SamSaffron** [comment]
❤️ 💟 😍  thanks heaps for working on this!

**Comment by cbliard** [comment]
Regarding the proposed changes, wouldn't it be easier and more performant to use `String.new`and keep using `<<` instead of replacing them with `+=`.

For instance from this piece of code:

```
body = "Rack Environment\n---------------\n"
env.each do |k,v|
  body << "#{k}: #{v}\n"
end
body << "\n\nEnvironment\n---------------\n"
text_result(body)
```

Instead of changing it to this:
```
body = "Rack Environment\n---------------\n"
env.each do |k,v|
  body += "#{k}: #{v}\n"
end
body += "\n\nEnvironment\n---------------\n"
text_result(body)
```

Use this:
```
body = String.new("Rack Environment\n---------------\n")
env.each do |k,v|
  body << "#{k}: #{v}\n"
end
body << "\n\nEnvironment\n---------------\n"
text_result(body)
```

**Comment by redtachyons** [comment]
I thought about the same, But in that case, we are keeping strings as mutable, right? 

> and more performant to use 

Can you please share the performance benchmarks comparing String.new and immutable string operations

**Comment by cbliard** [comment]
Exactly, some strings are kept as mutable and the majority are frozen. Instead of having all strings mutables by default and choosing which ones are frozen, the frozen string literal option makes all strings frozen by default and we choose which ones are mutable. Having mutable strings is not forbidden.

About benchmarking `+=` and `<<`, here is a little benchmark using the `benchmark-ips` gem

```
require 'benchmark/ips'

CHAR = 'A'.freeze
[10, 100, 1000, 10000].each do |count|
  puts "concatenating #{count} strings together"
  Benchmark.ips do |x|
    x.config(time: 3, warmup: 1)
    x.report("+=") do
      s = ""
      count.times { s += CHAR }
    end

    x.report("<<") do |times|
      s = String.new
      count.times { s << CHAR }
    end

    x.compare!
  end
end
```

It shows that `+=` is an order of magnitude slower than `<<`, which is logical because += allocates a new string everytime and that creates a lot of short-living objects.

I have not read mini-profiler code thoroughly and this may be a micro-optimization as it does not matter so much if there are only a couple of string concatenations.

**Comment by SamSaffron** [comment]
Huge warning on string.new, you get an ascii encoded string back, always
prefer “”.dup which is utf8

**Comment by cbliard** [comment]
Thanks, I did not know about that.

File: lib/patches/db/nobrainer.rb [hunk 1]
```diff
@@ -9,16 +10,16 @@ def on_query(env)
       query = ""
```
Comment (SamSaffron) [inline lib/patches/db/nobrainer.rb]: The general pattern should be ... 

```
query = "".dup
```

Then rest of the code is unchanged, all the string copying here should probably change

**Comment by SamSaffron** [comment]
@cbliard  agree, the general pattern is that any "mutable" string we have should be initialized with `"".dup` then later on we can still use `<<` like we do today.

File: lib/mini_profiler/timer_struct/client.rb [hunk 2]
```diff
@@ -16,10 +17,7 @@ def self.init_instrumentation
         # used by Railtie to instrument asset_tag for JS / CSS
         def self.instrument(name, orig)
           probe = "<script>mPt.probe('#{name}')</script>"
-          wrapped = probe
-          wrapped << orig
-          wrapped << probe
-          wrapped
+          probe + orig + probe + orig
```
Comment (cbliard) [inline lib/mini_profiler/timer_struct/client.rb]: Warning here, there is an extra `orig`. It should be `probe + orig + probe`.

Comment (redtachyons) [reply lib/mini_profiler/timer_struct/client.rb]: check the test here https://github.com/MiniProfiler/rack-mini-profiler/blob/master/spec/lib/timer_struct/client_timer_struct_spec.rb#L41

Comment (cbliard) [reply lib/mini_profiler/timer_struct/client.rb]: Indeed. Sorry for the burden.

**Pull request closed without merging**